proposals/0000-local-types.md

```diff
@@ -0,0 +1,438 @@
+[reverse dependencies of the `reflection` package](http://packdeps.haskellers.com/reverse/reflection).
+
+## Proposed Change Specification
+
+Allow type and instance declarations in `let` expressions and `where` clauses.
+
+### Examples
+
+Suppose we want to serialize and deserialize values differently depending on
+command-line arguments. Then we could write
+
+```haskell
+import System.Environment
+import Data.Aeson
+
+data B s = ...
+foo :: forall s. ... B s ... -> IO ...
+
+main = do
+  args <- getArgs
+  let
+    data A
+    instance FromJSON (B A) where
+      parseJSON = ... -- using args
+    instance ToJSON (B A) where
+      toJSON = ... -- again using args
+  foo @A ...
+```
+
+The `reflection` package defines
+
+```haskell
+class Reifies s a | s -> a where
+  -- | Recover a value inside a 'reify' context, given a proxy for its
+  -- reified type.
+  reflect :: proxy s -> a
+
+-- | Reify a value at the type level, to be recovered with 'reflect'.
+reify :: forall a r. a -> (forall (s :: *). Reifies s a => Proxy s -> r) -> r
+```
+
+We can implement `reify` with no difficulty whatsoever:
+
+```haskell
+reify :: forall a r. a -> (forall (s :: *). Reifies s a => Proxy s -> r) -> r
+reify a f = f (Proxy :: Proxy S)
+  where
+    data S
+    instance Reifies S a where
+      reflect _ = a
+```
+
+### Type declarations
+
+Local `data` and `newtype` declarations have forms similar to global ones,
+but their kind signatures may not mention local type variables, as Richard
+Eisenberg believes this would be a significant complication). Local type
+declarations are local to their scope and may not escape: all terms must
+be typed using only global types and local types in scope.
+
+```haskell
+-- OK: *values* involving local types may escape.
+data Hide = forall a. Hide a
+f :: Int -> Hide
+f x = Hide (P x)
+  where
+    newtype P = P Int
+
+-- Prohibited: the type P escapes its scope.
+g x = P x
+  where
+    newtype P = P Int
+```
+
+### Type synonyms
+
+Local type synonym definitions would be allowed, working exactly as they
+do elsewhere.
+
+### Class instances
+
+It is well known that allowing arbitrary local instances destroys coherence.
+One could write
+
+```haskell
+import Data.Constraint
+-- data Dict c where
+--   Dict :: c => Dict c
+
+f :: forall a. (a -> a -> a) -> Dict (Semigroup a)
+f append = Dict
+  where
+    instance Semigroup a where
+      (<>) = append
+```
+
+at which point matching on two applications of `f` could reveal conflicting
+`Semigroup` instances for the same type. We must not allow that!
+
+I believe the key to maintaining coherence simply is to reject *local* instances
+in favor of *locally defined* instances. All instances remain (conceptually)
+global, but they may involve local types and values. To keep things simple, we
+impose the following requirement: the instance head of a locally defined instance
+must mention a local type declared in the *same* `let` or `where`. We could
```

Comment:
We definitely want to reject that instance, indeed! But not, to my way of thinking, because the *instance* is determined by `Int`, but rather because the fundep indicates that `B` is determined by `Int`. We forbid `B` on the RHS of the fundep arrow because the LHS (`Int`) does not involve anything that lies within the scope of `B`.

Reply:
Yes, I think your phrasing is more correct. But as a consequence of your restriction, I think local instances still have to be determined by local types, unless you're doing something silly like a local instance for global types, e.g. `instance Foo Int Char`.

Reply:
I think they're both *correct*, but mine is more *syntactic* and therefore much easier to explain and check. We get to decide whether to allow silly instances like the one you indicate; the more I think about it the more I think they should produce warnings rather than errors.